### PR TITLE
dev/core#2396 - Read afform defn from .aff.php files

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Makes it possible to package Afforms in an extension with translatable page titles, descriptions, etc.

See https://lab.civicrm.org/dev/core/-/issues/2396

Before
---------
Not possible to translate Afform titles.

After
--------
Afform titles, descriptions, etc can be wrapped in `E::ts()`.

Technical Details
-------
This is intended for developers and extensions authors. They can generate these php files by calling Afform.get with the api explorer, and copying the result (minus the "layout" field).
FormBuilder still saves files locally in `aff.json` format. Overriding a packaged `.aff.php` form with a local `aff.json` works fine.

### CODE REVIEW

**Comment by totten** [comment]
> This is intended for developers and extensions authors. They can generate these php files by calling Afform.get with the api explorer, and copying the result (minus the "layout" field).
> FormBuilder still saves files locally in aff.json format. Overriding a packaged .aff.php form with a local aff.json works fine.

This is a clever way to skirt the issue of read-write support for `.aff.php` in FormBuilder. AFAICS, the main issue is that any local overrides will be lossy, e.g.

* Dev creates `afsearchAdministerLocationTypes.aff.php` with `ts('Location Types')`
* Local admin edits `afsearchAdministerLocationTypes`. (For example, they replace the long help-blurb with something simpler.) System saves the updated `afsearchAdministerLocationTypes.aff.json` and `afsearchAdministerLocationTypes.aff.html`.
* Effect: We lose the `ts()` call. The content of `title` will no longer be translated. (It's locked into whatever language was being used by the local-admin.) 

In favor of this patch:

* The lossy-save issue isn't technically regressive. (It's a new bug, and it's less impactful than the old bug being fixed...)
* Subsequent work on read-write for `*.aff.php` would probably use a lot of this. (Maybe 80-100% of these changes would still be needed.)

I guess the main question is whether we get boxed-in to extra work by shipping the intermediate form. (Ex: 5.67 ships with the patch. When site-builders do customizations, they do some lossy saves. In the future 5.69, do we need a data-cleanup/migration-path?) But I think the cleanup/migration issue arises either way. Downstream already has JSON metadata without `ts()` expressions (in `[civicrm.files]` and in exts). I suppose the migration question comes up regardless of which format we introduce and how it's timed.

Also in favor this patch: it allows us to try the `*.aff.php` / `ts()` format for a bit in some isolated contexts before we undertake full conversion...

**Comment by colemanw** [comment]
@totten yea I think any future improvements could build on this patch. Ideas: 

- **Progressive overrides** so that the local `aff.json` file doesn't contain the complete definition; only stuff that's different from the base definition.
- **Multilingual support** instead of just a `title` field, the definition file can have `title-en_US`, `title-es_ES` etc.

I don't see this PR as boxing us in; both of those ideas could work with either file format.

**Comment by mlutfy** [comment]
>  Afform titles, descriptions, etc can be wrapped in ts().

`E::ts`, right?

> Makes it possible to package Afforms in an extension with translatable page titles, descriptions, etc.

Could you provide an example of how the extension would package the titles/descriptions? Is it something devs have to do manually, or is it automatically generated?

File: ext/civicrm_admin_ui/ang/afsearchAdminContactTypes.aff.php [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?php
+use CRM_CivicrmAdminUi_ExtensionUtil as E;
+
+return [
+  'type' => 'search',
+  'title' => E::ts('Contact Types'),
+  'description' => E::ts('Administer contact types and sub-types'),
```
Comment (colemanw) [inline ext/civicrm_admin_ui/ang/afsearchAdminContactTypes.aff.php:7]: @mlutfy it's just php files with translatable strings, like this ^

Comment (colemanw) [inline ext/civicrm_admin_ui/ang/afsearchAdminContactTypes.aff.php:7]: @mlutfy it's just php files with translatable strings, like this ^

**Comment by colemanw** [comment]
retest this please

**Comment by totten** [comment]
Once more unto the breach, dear friends, once more!

**Pull request merged**